src/openforms/js/components/admin/form_design/variables/utils.js

```diff
@@ -141,20 +142,42 @@ const updateFormVariables = (
       });
     }
 
-    // When a component is removed, oldComponent is null
-    updatedFormVariables = updatedFormVariables.filter(
-      variable => !keysToRemove.includes(variable.key)
-    );
+    updatedFormVariables = updatedFormVariables.filter(variable => {
+      const matchKeyToRemove = keysToRemove.includes(variable.key);
+
+      // In the case that there are duplicate keys, we need to figure out which of the variables with duplicate keys
+      // should be removed. Since in a step there can't be duplicate keys, check that the formDefinition matches
+      return !matchKeyToRemove || variable.formDefinition !== formDefinition;
+    });
   }
 
   return updatedFormVariables;
 };
 
+const checkForDuplicateKeys = (formVariables, staticVariables) => {
+  let validationErrors = [];
+  let existingKeys = [];
+
+  staticVariables.concat(formVariables).map((variable, index) => {
+    if (existingKeys.includes(variable.key)) {
+      validationErrors.push([`variables.${index}.key`, 'unique']);
+
+      if (!variable.errors) variable.errors = {};
+      variable.errors['key'] = 'unique';
```

Comment:
maybe we can skip the whole constant for error messages, and use `defineMessage` directly here? I don't think we need this layer of indirection, and it makes it possible to cater the exact error message depending on the situation (component vs. user defined variable).